### PR TITLE
Arrow upload

### PR DESCRIPTION
**Pull request by bmschmidt**
Changelog for users.
1. Uploads are now internally handled as Arrow tables, allowing greater type safety and data throughput.
2. In addition to passing lists of dicts, you can directly pass pandas dataframes or pyarrow tables to any upload methods.
3. Datetime formats are now passed as native python dates or datetimes (or as pandas date or datetime). ISO-formatted strings will no longer be automatically coerced--just pass your own.
4. Null values are now allowed in any fields except for embeddings and ids. These can be passed either by setting the key to None, omiting a key from a dictionary, or using a pandas null type. 
5. Typechecking is stricter than before, with the aim of raising errors on the client side sooner.

### CODE REVIEW

File: nomic/project.py [hunk 1]
```diff
@@ -163,99 +175,81 @@ def _get_index_job(self, job_id: str):
 
         return response.json()
 
-    def _validate_and_correct_user_supplied_metadata(
-        self, data: List[Dict], project, replace_empty_string_values_with_string_null=True
-    ):
+    def _validate_and_correct_arrow_upload(
+        self, data: pa.Table, project: "AtlasProject"
```
Comment (AndriyMulyar) [inline nomic/project.py:179]: We need to test that empty strings to do break things or cause strange behavior on search

File: nomic/project.py [hunk 2]
```diff
@@ -163,99 +175,81 @@ def _get_index_job(self, job_id: str):
...
-                    if math.isnan(datum[key]):
-                        raise Exception(
-                            f"Metadata sent to Atlas must be a flat dictionary. Values must be strings, floats or ints. Key `{key}` of datum {str(datum)} is in violation."
-                        )
-
-                if not isinstance(datum[key], (str, float, int)):
-                    raise Exception(
-                        f"Metadata sent to Atlas must be a flat dictionary. Values must be strings, floats or ints. Key `{key}` of datum {str(datum)} is in violation."
-                    )
+        if project.id_field == ATLAS_DEFAULT_ID_FIELD and not ATLAS_DEFAULT_ID_FIELD in data.column_names:
```
Comment (AndriyMulyar) [inline nomic/project.py:205]: nice

File: nomic/project.py [hunk 3]
```diff
@@ -1354,228 +1360,148 @@ def delete_data(self, ids: List[str]) -> bool:
...
-
-        upload_endpoint = "/v1/project/data/add/json/initial"
-        if progressive:
-            upload_endpoint = "/v1/project/data/add/json/progressive"
+    ):
+        """
+        Add text data to the project.
+        data: A pandas DataFrame, list of dictionaries, or pyarrow Table matching the project schema.
+        pbar: (Optional). A tqdm progress bar to display progress.
+        shard_size: (Deprecated).
```
Comment (AndriyMulyar) [inline nomic/project.py:1372]: Do not say deprecated, just remove from docstrings

File: nomic/project.py [hunk 4]
```diff
@@ -1354,228 +1360,148 @@ def delete_data(self, ids: List[str]) -> bool:
...
-            if get_object_size_in_bytes(data_shard) > 8000000:
-                raise Exception(
-                    "Your metadata upload shards are to large. Try decreasing the shard size or removing un-needed fields from the metadata."
+            data_shard = data.slice(i, shard_size)
+            with io.BytesIO() as buffer:
+                data_shard = data_shard.replace_schema_metadata({'project_id': self.id})
+                feather.write_feather(data_shard, buffer, compression = 'zstd', compression_level = 6)
+                buffer.seek(0)
+
+                # TODO: Actually put data in.
```
Comment (AndriyMulyar) [inline nomic/project.py:1490]: is to todo still necessary

**Review by AndriyMulyar — CHANGES_REQUESTED** [review]
